### PR TITLE
JIT: retype some return expressions

### LINKED ISSUE #21051: AF in in System.Globalization.CharUnicodeInfo:InternalGetNumericValue
Found by https://github.com/dotnet/coreclr/pull/21048. Remove NoOptimization workaround and run crossgen System.Private.CoreLib.dll.

Assert failure(PID 2976 [0x00000ba0], Thread: 3816 [0x0ee8]): Assertion failed '((regMask & emitThisGCrefRegs) && (ins == INS_add)) || ((regMask & emitThisByrefRegs) && (ins == INS_add || ins == INS_sub))' in 'System.Globalization.CharUnicodeInfo:InternalGetNumericValue(int):double' (IL size 175)

### PR DESCRIPTION
**Pull request by AndyAyersMS**
The jit is fairly tolerant of byref/native int mismatches for inline arguments
and return values. And some of the new unsafe helpers do this kind of
reinterpretation across call boundaries as well. This sometimes leads the jit
to lose track of byrefs.

A general fix for this is likely somewhat involved. For now we simply detect if
we're about to lose track of a byref when substituting a particular kind of
return expression, and retype the expression to a byref.

Fixes #21051.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@briansull PTAL
cc @dotnet/jit-contrib 

Note this is a targeted workaround for this specific issue.

No diffs (w/ same corelib)

File: src/System.Private.CoreLib/shared/System/Globalization/CharUnicodeInfo.cs [hunk 1]
```diff
@@ -297,7 +296,6 @@ public static UnicodeCategory GetUnicodeCategory(int codePoint)
         ////////////////////////////////////////////////////////////////////////
 
         // Workaround for https://github.com/dotnet/coreclr/issues/21051
```
Comment (benaadams) [inline src/System.Private.CoreLib/shared/System/Globalization/CharUnicodeInfo.cs]: Also delete the "Workaround" comments?

Comment (AndyAyersMS) [reply src/System.Private.CoreLib/shared/System/Globalization/CharUnicodeInfo.cs]: Yes. Thanks for spotting that.

**Comment by AndyAyersMS** [comment]
Might have restarted a bit too soon.

@dotnet-bot test this please

**Comment by AndyAyersMS** [comment]
[CentOS](https://ci.dot.net/job/dotnet_coreclr/job/master/job/x64_checked_centos7.1_innerloop_flow_prtest/9757/) download failure:
```
/mnt/resource/j/workspace/dotnet_coreclr/master/x64_checked_centos7.1_innerloop_prtest/Tools/dotnetcli/sdk/2.1.301/NuGet.targets(114,5): 
error : Failed to download package 'System.Reflection.Extensions.4.0.0' from 'https://dotnetfeed.blob.core.windows.net/dotnet-core/flatcontainer/system.reflection.extensions/4.0.0/system.reflection.extensions.4.0.0.nupkg'. 
```
[CoreFX tests](https://ci.dot.net/job/dotnet_coreclr/job/master/job/x64_release_windows_nt_corefx_innerloop_prtest/4124/) also seems to have run into a download failure:
```
23:56:09 Error downloading tests from: https://cloudcijobs.blob.core.windows.net/coreclrci/CoreFXArchives/TestList.json
23:56:09 An error occurred while sending the request.
23:56:09 RUNTEST: CoreFX tests finished with failures. Refer to the following:
```
[OSX](https://ci.dot.net/job/dotnet_coreclr/job/master/job/x64_checked_osx10.12_innerloop_flow_prtest/10454/) disconnected while building tests:
```
00:57:17 FATAL: command execution failed
00:57:17 java.nio.channels.ClosedChannelException
00:57:17 	at org.jenkinsci.remoting.protocol.impl.ChannelApplicationLayer.onReadClosed(ChannelApplicationLayer.java:208)
```

**Comment by AndyAyersMS** [comment]
@dotnet-bot test Windows_NT x64 Release CoreFX Tests
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test
@dotnet-bot test CentOS7.1 x64 Checked Innerloop Build and Test

**Pull request merged**